lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -0,0 +1,261 @@
+    Value distributedVal = warpOp.getResult(yieldOperand->second);
+    SmallVector<Value> yieldValues;
+    SmallVector<Type> retTypes;
+    SmallVector<unsigned> operandForwarded;
+    for (OpOperand &operand : elementWise->getOpOperands()) {
+      auto targetType = VectorType::get(
+          distributedVal.getType().cast<VectorType>().getShape(),
+          operand.get().getType().cast<VectorType>().getElementType());
+      operandForwarded.push_back(operand.getOperandNumber());
+      retTypes.push_back(targetType);
+      yieldValues.push_back(operand.get());
+    }
+    WarpSingleLaneOp newWarpOp =
+        moveRegionToNewWarpAddReturn(rewriter, warpOp, yieldValues, retTypes);
+    SmallVector<Value> newOperands(elementWise->getOperands().begin(),
+                                   elementWise->getOperands().end());
+    for (unsigned i : operandForwarded) {
+      newOperands[i] = newWarpOp.getResult(i + warpOp.getNumResults());
+    }
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointAfter(newWarpOp);
+    Operation *newOp = cloneOpWithOperandsAndTypes(
+        rewriter, warpOp.getLoc(), elementWise, newOperands,
+        {warpOp.getResult(yieldOperand->second).getType()});
+    newWarpOp.getResult(yieldOperand->second)
+        .replaceAllUsesWith(newOp->getResult(0));
+    rewriter.eraseOp(warpOp);
+    return success();
+  }
+};
+
+/// Sink out transfer_read op feeding into a warp op yield.
+/// ```
+/// %0 = vector_ext.warp_execute_on_lane_0(%arg0) -> (vector<1xf32>) {
+///   ...
+//    %2 = vector.transfer_read %src[%c0], %cst : memref<1024xf32>,
+//    vector<32xf32>
+///   vector_ext.yield %2 : vector<32xf32>
+/// }
+/// ```
+/// To
+/// ```
+/// %dead = vector_ext.warp_execute_on_lane_0(%arg0) -> (vector<1xf32>,
+/// vector<1xf32>, vector<1xf32>) {
+///   ...
+///   %2 = vector.transfer_read %src[%c0], %cst : memref<1024xf32>,
+///   vector<32xf32> vector_ext.yield %2 : vector<32xf32>
+/// }
+/// %0 = vector.transfer_read %src[%c0], %cst : memref<1024xf32>, vector<1xf32>
+struct WarpOpTransferRead : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    auto operand = getWarpResult(
+        warpOp, [](Operation *op) { return isa<vector::TransferReadOp>(op); });
+    if (!operand)
+      return failure();
+    auto read = cast<vector::TransferReadOp>(operand->first);
+    Value distributedVal = warpOp.getResult(operand->second);
+
+    SmallVector<Value, 4> indices(read.indices().begin(), read.indices().end());
+    AffineMap map = calculateImplicitMap(read.getResult(), distributedVal);
+    AffineMap indexMap = map.compose(read.permutation_map());
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(warpOp);
+    for (auto it : llvm::zip(indexMap.getResults(), map.getResults())) {
```

Comment:
I think most of this loop goes away if you embed the multiplicity in the affine map ?

Reply:
SG, but please don't let it hanging, it will be significantly better if you can make it work.